hail/src/main/scala/is/hail/expr/ir/lowering/RVDToTableStage.scala

```diff
@@ -1,15 +1,92 @@
 package is.hail.expr.ir.lowering
 
-import is.hail.expr.ir.{IR, PartitionRVDReader, ReadPartition, StreamRange}
-import is.hail.rvd.RVD
+import java.io.{ByteArrayInputStream, ByteArrayOutputStream}
+
+import is.hail.annotations.{BroadcastRow, Region, RegionValue}
+import is.hail.asm4s._
+import is.hail.expr.ir.{Compile, CompileIterator, ExecuteContext, GetField, IR, In, Let, MakeStruct, PartitionRVDReader, ReadPartition, StreamArgType, StreamRange, ToArray}
+import is.hail.io.{BufferSpec, TypedCodecSpec}
+import is.hail.rvd.{RVD, RVDContext, RVDType}
+import is.hail.sparkextras.ContextRDD
+import is.hail.types.physical.{PArray, PStruct}
+import is.hail.utils.{FastIndexedSeq, FastSeq, box}
 
 object RVDToTableStage {
   def apply(rvd: RVD, globals: IR): TableStage = {
     TableStage(
       globals = globals,
       partitioner = rvd.partitioner,
+      dependency = TableStageDependency.fromRVD(rvd),
       contexts = StreamRange(0, rvd.getNumPartitions, 1),
       body = ReadPartition(_, rvd.rowType, PartitionRVDReader(rvd))
     )
   }
+}
+
+object TableStageToRVD {
+  def apply(ctx: ExecuteContext, ts: TableStage, relationalBindings: Map[String, IR]): (BroadcastRow, RVD) = {
+
+    val sparkContext = ctx.backend
+      .asSpark("TableStageToRVD")
+      .sc
+
+    val baseStruct = MakeStruct(FastSeq(
+      ("globals", ts.globals),
+      ("broadcastVals", MakeStruct(ts.broadcastVals)),
+      ("contexts", ToArray(ts.contexts))))
+    val globalsAndBroadcastVals = LowerToCDA.substLets(ts.letBindings.foldRight[IR](baseStruct) { case ((name, value), acc) =>
+      Let(name, value, acc)
+    }, relationalBindings)
+
+    val (gbPType: PStruct, f) = Compile[AsmFunction1RegionLong](ctx, FastIndexedSeq(), FastIndexedSeq(classInfo[Region]), LongInfo, globalsAndBroadcastVals)
+    val gbAddr = f(0, ctx.r)(ctx.r)
+
+    val globPType = gbPType.fieldType("globals").asInstanceOf[PStruct]
+    val globRow = BroadcastRow(ctx, RegionValue(ctx.r, gbPType.loadField(gbAddr, 0)), globPType)
+
+    val bcValsPType = gbPType.fieldType("broadcastVals")
+
+    val bcValsSpec = TypedCodecSpec(bcValsPType, BufferSpec.wireSpec)
+    val encodedBcVals = sparkContext.broadcast(bcValsSpec.encodeValue(ctx, bcValsPType, gbPType.loadField(gbAddr, 1)))
+    val (decodedBcValsPType: PStruct, makeBcDec) = bcValsSpec.buildDecoder(ctx, bcValsPType.virtualType)
+
+    val contextsPType = gbPType.fieldType("contexts").asInstanceOf[PArray]
+    val contextPType = contextsPType.elementType
+    val contextSpec = TypedCodecSpec(contextPType, BufferSpec.wireSpec)
+    val contextsAddr = gbPType.loadField(gbAddr, 2)
+    val nContexts = contextsPType.loadLength(contextsAddr)
+
+    val (decodedContextPType: PStruct, makeContextDec) = contextSpec.buildDecoder(ctx, contextPType.virtualType)
```

Comment:
Somehow `buildDecoder` is actually returning a `PArray`, and a test is failing. Maybe assert `contextPType` is a `PStruct` and figure out what is going wrong? I thought it was just a `contextPType` vs `contextsPType` mixup (the names are maybe too similar for easy reading) but after reading I don't see any instances of that.

Reply:
oh, thanks for the guidance!

Reply:
ah, I see what's going on -- contexts don't actually need to be structs! So I'll just have to wrap in a struct so the interpreted code (called from staged code) works out.

Reply:
Makes sense 👍